api/billingd/service/service.go

```diff
@@ -740,241 +717,135 @@ func (s *Service) handleStoredData(
-		usage.Total,
-		usage.Free,
-	)
-	return &pb.IncInstanceWritesResponse{
-		Period:         periodToPb(cus.Period),
-		InstanceWrites: usage,
-	}, nil
-}
-
-func (s *Service) getCustomer(ctx context.Context, key, val string) (*Customer, error) {
-	r := s.db.FindOne(ctx, bson.M{key: val})
-	if r.Err() != nil {
-		return nil, r.Err()
+	if err := cursor.Err(); err != nil {
+		return fmt.Errorf("cursor error: %v", err)
 	}
-	var doc Customer
-	return &doc, r.Decode(&doc)
+	return nil
 }
 
-func (s *Service) handleUsage(
-	ctx context.Context,
-	key string,
-	usage Usage,
-	usageKey string,
-	inc, unitSize, freeUnits int64,
-	billable bool,
-) (*pb.Usage, error) {
-	total := usage.Total + inc
-	if total < 0 {
-		total = 0
-	}
-	update := bson.M{usageKey + ".total": total}
-	units := int64(math.Round(float64(total) / float64(unitSize)))
-	if units > freeUnits && !billable {
-		return nil, common.ErrExceedsFreeUnits
+func (s *Service) reportDailyUnits(product Product, usage Usage) error {
+	var freeSize, paidSize int64
+	if usage.Total > product.FreeQuotaPerInterval {
+		freeSize = product.FreeQuotaPerInterval
+		paidSize = usage.Total - product.FreeQuotaPerInterval
+	} else {
+		freeSize = usage.Total
 	}
-	if units != usage.Units {
+	freeUnits := getDailyUnits(product, freeSize)
+	if freeUnits > 0 {
 		if _, err := s.stripe.UsageRecords.New(&stripe.UsageRecordParams{
-			SubscriptionItem: stripe.String(usage.ItemID),
-			Quantity:         stripe.Int64(units),
+			SubscriptionItem: stripe.String(usage.FreeItemID),
+			Quantity:         stripe.Int64(freeUnits),
 			Timestamp:        stripe.Int64(time.Now().Unix()),
-			Action:           stripe.String(stripe.UsageRecordActionSet),
+			Action:           stripe.String(stripe.UsageRecordActionIncrement),
 		}); err != nil {
-			return nil, err
+			return err
 		}
-		update[usageKey+".units"] = units
-	}
-	if _, err := s.db.UpdateOne(ctx, bson.M{"_id": key}, bson.M{"$set": update}); err != nil {
-		return nil, err
 	}
-	free := (freeUnits * unitSize) - total
-	if free < 0 {
-		free = 0
+	paidUnits := getDailyUnits(product, paidSize)
+	if paidUnits > 0 {
+		if _, err := s.stripe.UsageRecords.New(&stripe.UsageRecordParams{
+			SubscriptionItem: stripe.String(usage.PaidItemID),
+			Quantity:         stripe.Int64(paidUnits),
+			Timestamp:        stripe.Int64(time.Now().Unix()),
+			Action:           stripe.String(stripe.UsageRecordActionIncrement),
+		}); err != nil {
+			return err
+		}
 	}
-	return &pb.Usage{
-		Units: units,
-		Total: total,
-		Free:  free,
-	}, nil
+	return nil
+}
+
+func getDailyUnits(product Product, size int64) int64 {
+	units := float64(size) / float64(product.UnitSize)
+	return int64(math.Round(units))
+}
+
+func getPeriodDays() int {
+	// Go to next month
+	d1 := time.Now().AddDate(0, 1, 0)
+	// Zeroing days will cause time to normalize the date to the last day of the current month.
+	d2 := time.Date(d1.Year(), d1.Month(), 0, 0, 0, 0, 0, time.UTC)
+	return d2.Day()
 }
```

Comment:
If we're trying to get the last day of the month, I'm not sure this is correct for a border case: last day of the month of a 31 day month.

Easier to explain [here](https://play.golang.org/p/V_S7EXrCt6a).
There also I propose another solution which I think doesn't have this problem, since day is normalized to day 1 before making the hairy 1-month-sum, so avoiding the 30 vs 31 prob.

Reply:
Ah, good catch. Though I'm not sure the nanosecond addition is needed. This also works:

```
func getPeriodDays(now time.Time) int {
	// Go to next month
	d1 := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, time.UTC).AddDate(0, 1, 0)
	// Zeroing days will cause time to normalize the date to the last day of the current month.
	d2 := time.Date(d1.Year(), d1.Month(), 0, 0, 0, 0, 0, time.UTC)
	return d2.Day()
}
```

Though, I'm not actually using that method since it's not possible to dynamically set the unit price when reporting to stripe. So, kinda wrote that and forgot about it. Will remove.

Currently, monthly-type pricing is determined using 30 days for a month. However, I just had a thought... why not use the average of 30.4167.

Reply:
Yeah, I think assuming 30 is fine if 30.4167 might make some manual-calculations unnecessarily hard (if any manual calc makes sense).

Reply:
Stuck with the average for now.